File: Python3/1 - 200/17 Letter Combinations of a Phone Number.py
# 40ms, 100.0%
class Solution:
    def letterCombinations(self, digits):
        def generator(digits):
            if len(digits) == 1:
                for ch in letters[ord(digits[0]) - ord('0')]:
                    yield (ch,)
            else:
                for ch in letters[ord(digits[0]) - ord('0')]:
                    the_rest = generator(digits[1:])
                    for tr in the_rest:
                        yield (ch,) + tr
        
        if not digits: return []
        letters = ['', '', 'abc', 'def', 'ghi', 'jkl',
                   'mno', 'pqrs', 'tuv', 'wxyz']
        return [''.join(word) for word in generator(digits)]
    

File: 1 - 200/17 Letter Combinations of a Phone Number.cc
// 3ms, 0.78%
class Solution
{
public:
    vector<string> letterCombinations(string digits)
    {
        if(digits.empty()) return vector<string>();
        
        static const vector<string> v = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
        vector<string> result;
        result.push_back("");
        
        for(int i = 0 ; i < digits.size(); ++i)
        {
            int num = digits[i] - '0';
            if(num < 0 || num > 9) break;
            if(v[num].empty()) continue;
            
            const string &candidate = v[num];
            vector<string> tmp;
            for(int j = 0 ; j < candidate.size() ; ++j)
                for(int k = 0 ; k < result.size() ; ++k)
                    tmp.push_back(result[k] + candidate[j]);
            result.swap(tmp);
        }
        return result;
    }
};
